libsignal-service-actix/src/push_service.rs

```diff
@@ -182,6 +183,70 @@ impl PushService for AwcPushService {
         ))
     }
 
+    async fn post_to_cdn0<'s, C: std::io::Read + Send + 's>(
+        &mut self,
+        path: &str,
+        value: &[(&str, &str)],
+        file: Option<(&str, &'s mut C)>,
+    ) -> Result<(), ServiceError> {
+        let url = Url::parse(&self.cfg.cdn_urls[&0])
+            .expect("valid cdn base url")
+            .join(path)
+            .expect("valid CDN path");
+
+        log::debug!("AwcPushService::post_to_cdn({:?})", url);
+        let mut client = self.client.post(url.as_str());
+
+        let mut form = multipart::Form::default();
+
+        for (k, v) in value {
+            form.add_text(*k, *v);
+        }
+
+        if let Some((filename, file)) = file {
+            // XXX Actix doesn't cope with none-'static lifetimes
+            // https://docs.rs/actix-web/3.2.0/actix_web/body/enum.Body.html
+            let mut buf = Vec::new();
+            file.read_to_end(&mut buf)
+                .expect("infallible Read instance");
+            form.add_reader(filename, std::io::Cursor::new(buf));
+            client = client.header("Content-Type", "application/octet-stream");
+        }
+
+        let content_type = form.content_type();
+
+        // XXX Amazon S3 needs the Content-Length, but we don't know it without depleting the whole
+        // stream.  Sadly, Content-Length != contents.len(), but should include the whole form.
```

Comment:
Just being curious here: how's `Content-Length` is not `contents.len()`? There _must_ be something we can do to calculate the right value... but I guess you already tried that :smile:

Reply:
Because `contents` contains the attachment, and what you send to the Amazon S3 server is a form (yes, an HTML form), that contains a bunch of field *and* the attachment. As such, it depends on how the serialization of the form happens, which is why I serialize before (to memory), and count, and then send.

AttachmentV3 does not have that issue, by the way, but is way more involved (has a few extra requests). It has as advantage that you can stream, without knowing contents-length, and resume an upload later on.

Reply:
Puh. Ok, I accept this (foul) explanation :smiley: